spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql/src/main/java/org/springframework/cloud/gcp/sql/CloudSqlJdbcUrlProvider.java

```diff
@@ -0,0 +1,29 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.sql;
+
+/**
+ * Implementations of this interface typically construct a JDBC URL for Cloud SQL from a GCP
+ * project ID and an instance name.
+ *
+ * @author João André Martins
+ */
+@FunctionalInterface
+public interface CloudSqlJdbcUrlProvider {
```

Comment:
We thought about it, but I don't think we can because there are several fields that are specific to our DataSource configuration: project ID, region name and instance name. Project ID is already in `spring.cloud.gcp`, but we needed another namespace to create the other two (`spring.cloud.gcp.sql`). Also using `DataSourceProperties` would mix namespaces, and this could be confusing to users.

We could use `DataSourceProperties` as a POJO to feed into `cloudSqlDataSource()`, and then just do the field mapping inside the method, but I don't see many advantages in this -- the POJO would just do to carry the URL since the username and password are already in the `GcpCloudSqlProperties` object.